R/test-DataClass.R

```diff
@@ -121,6 +121,49 @@ test_download <- function(DataClass_obj, download, snapshot_path) {
   }
 }
 
+
+#' Test download method for JSON files works correctly
+#' @description Test data can be downloaded if `download = TRUE`, or a requested
+#' snapshot file is not found, and store a snap shot in the `snapshot_dir`. If
+#' an existing snapshot file is found then load this data to use in future tests
+#' @param DataClass_obj The R6Class object to perform checks on.
+#' Must be a `DataClass` or `DataClass` child object.
+#' @param download Logical check to download or use a snapshot of the data
+#' @param snapshot_path character_array the path to save the downloaded
+#' snapshot to.
+#' @importFrom purrr map walk
+#' @importFrom dplyr slice_head
+#' @family tests
+#' @concept tests
+#' @export
+test_download_JSON <- function(DataClass_obj, download, snapshot_path) {
+  if (!file.exists(snapshot_path)) {
+    download <- TRUE
+  }
+  if (download) {
+    testthat::test_that(
+      paste0(DataClass_obj$data_name, " downloads sucessfully"),
+      {
+        DataClass_obj$download_JSON()
+        walk(DataClass_obj$data$raw, function(data) {
+          testthat::expect_s3_class(data, "data.frame")
+          testthat::expect_true(nrow(data) > 0)
```

Comment:
Some of these tests (which are copied straight over from the CSV code) sort of implicitly assume that we are getting table-like inputs from JSON but even with the Vietnam data the clean method is having to do more “heavy lifting”, and that’s just because we’re getting three parallel lists of tables (roughly) with a list to tie them together. Conceivably we could have other JSON input which was even further from our rectangular array assumptions but still valid JSON, useful data, and convertible (by cleaning) into something the rest of our class can work with.

I don’t necessarily think this needs to be changed now but I could imagine a case where we have another input which generates valid raw data which fails this test - and I’ve already made the argument about why cleaning should be done in the clean methods so I’m inclined to keep the download functions from starting to squeeze data into our rectangular table straitjacket.

Reply:
Yes very much agree. Hmm that does leave us in a bit of a poser then regarding what to test. I suppose but lack of error and the existence of a JSON object as a minimum?